File: fsharp/tracks-on-tracks-on-tracks/TracksOnTracksOnTracks.fs
module TracksOnTracksOnTracks

let newList: string list = []

let existingList: string list = ["F#"; "Clojure"; "Haskell"]

let addLanguage (language: string) (languages: string list): string list =
    language :: languages

let countLanguages (languages: string list): int = List.length languages

let reverseList(languages: string list): string list = List.rev languages 

let excitingList (languages: string list): bool =
    match languages with
    | "F#" :: _ -> true
    | _ :: tail ->
        match tail with
        | "F#" :: _ -> languages.Length = 2 || languages.Length = 3
        | _ -> false
    |  _ -> false
    

File: csharp/tracks-on-tracks-on-tracks/TracksOnTracksOnTracks.cs
using System;
using System.Collections.Generic;

public static class Languages
{
    public static List<string> NewList() => new();

    public static List<string> GetExistingLanguages() 
        => new() {"C#", "Clojure", "Elm"};

    public static List<string> AddLanguage(List<string> languages, string language)
    {
        languages.Add(language);
        return languages;
    }

    public static int CountLanguages(List<string> languages) => languages.Count;

    public static bool HasLanguage(List<string> languages, string language)
        => languages.Contains(language);

    public static List<string> ReverseList(List<string> languages)
    {
        languages.Reverse();
        return languages;
    }

    public static bool IsExciting(List<string> languages)
         => (languages.Count > 0 && languages[0] == "C#") || 
            (languages.Count is > 1 and <= 3 && languages[1] == "C#");

    public static List<string> RemoveLanguage(List<string> languages, string language)
    {
        languages.Remove(language);
        return languages;
    }

    public static bool IsUnique(List<string> languages)
    {
        var langSet = new HashSet<string>(languages);
        return languages.Count == langSet.Count;
    }
}
